### PR TITLE
feat: drop based on LRU (#2075)

### LINKED ISSUE #2075: LRU Chunk Retention Not Working
As reported by @e-dard 

```
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
| id | partition_key       | table_name | storage           | lifecycle_action | memory_bytes | object_store_bytes | row_count | time_of_first_write           | time_of_last_write            | time_closed |
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
| 46 | 2021-07-20 19:00:00 | query_log  | OpenMutableBuffer |                  | 26384621     |                    | 81467     | 2021-07-20 19:36:15.137247762 | 2021-07-20 19:37:07.183993002 |             |
| 0  | 2021-07-20 03:00:00 | query_log  | OpenMutableBuffer |                  | 1416333841   |                    | 125       | 2021-07-20 18:35:43.784786979 | 2021-07-20 19:37:04.318117465 |             |
| 3  | 2021-07-20 12:00:00 | query_log  | ReadBuffer        |                  | 37372        |                    | 4         |                               |                               |             |
| 3  | 2021-07-20 16:00:00 | query_log  | ReadBuffer        |                  | 29555        |                    | 2         |                               |                               |             |
| 1  | 2021-07-20 06:00:00 | query_log  | ReadBuffer        |                  | 22121        |                    | 1         |                               |                               |             |
| 36 | 2021-07-20 18:00:00 | query_log  | ReadBuffer        |                  | 22128        |                    | 1         |                               |                               |             |
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
| id | partition_key       | table_name | storage           | lifecycle_action | memory_bytes | object_store_bytes | row_count | time_of_first_write           | time_of_last_write            | time_closed |
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
| 46 | 2021-07-20 19:00:00 | query_log  | OpenMutableBuffer |                  | 27580581     |                    | 85403     | 2021-07-20 19:36:15.137247762 | 2021-07-20 19:37:12.564318929 |             |
| 0  | 2021-07-20 03:00:00 | query_log  | OpenMutableBuffer |                  | 1416333841   |                    | 125       | 2021-07-20 18:35:43.784786979 | 2021-07-20 19:37:04.318117465 |             |
| 3  | 2021-07-20 12:00:00 | query_log  | ReadBuffer        |                  | 37372        |                    | 4         |                               |                               |             |
| 3  | 2021-07-20 16:00:00 | query_log  | ReadBuffer        |                  | 29555        |                    | 2         |                               |                               |             |
| 1  | 2021-07-20 06:00:00 | query_log  | ReadBuffer        |                  | 22121        |                    | 1         |                               |                               |             |
| 36 | 2021-07-20 18:00:00 | query_log  | ReadBuffer        |                  | 22128        |                    | 1         |                               |                               |             |
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
| id | partition_key       | table_name | storage           | lifecycle_action | memory_bytes | object_store_bytes | row_count | time_of_first_write           | time_of_last_write            | time_closed |
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
| 47 | 2021-07-20 19:00:00 | query_log  | OpenMutableBuffer |                  | 437992       |                    | 1507      | 2021-07-20 19:37:16.624055193 | 2021-07-20 19:37:17.810653827 |             |
| 0  | 2021-07-20 03:00:00 | query_log  | OpenMutableBuffer |                  | 1416333841   |                    | 125       | 2021-07-20 18:35:43.784786979 | 2021-07-20 19:37:04.318117465 |             |
| 3  | 2021-07-20 12:00:00 | query_log  | ReadBuffer        |                  | 37372        |                    | 4         |                               |                               |             |
| 3  | 2021-07-20 16:00:00 | query_log  | ReadBuffer        |                  | 29555        |                    | 2         |                               |                               |             |
| 1  | 2021-07-20 06:00:00 | query_log  | ReadBuffer        |                  | 22121        |                    | 1         |                               |                               |             |
| 36 | 2021-07-20 18:00:00 | query_log  | ReadBuffer        |                  | 22128        |                    | 1         |                               |                               |             |
+----+---------------------+------------+-------------------+------------------+--------------+--------------------+-----------+-------------------------------+-------------------------------+-------------+
```

The retention policy is incorrectly dropping recent chunks before earlier chunks. This likely relates to recent changes to time_of_first_write and time_of_last_write as part of #1927. We should reintroduce functionality to the CatalogChunk to allow it to expose the necessary data for use by an LRU policy

### PR DESCRIPTION
**Pull request by tustvold**
Closes #2075

Uses the access timestamps added by #2081 to drive the lifecycle's drop policy

### CODE REVIEW

File: server/src/db/lifecycle.rs [hunk 1]
```diff
@@ -291,8 +292,24 @@ impl LifecycleChunk for CatalogChunk {
             .expect("failed to clear lifecycle action")
     }
 
-    fn time_of_first_write(&self) -> Option<DateTime<Utc>> {
-        self.time_of_first_write()
+    fn min_timestamp(&self) -> DateTime<Utc> {
```
Comment (tustvold) [inline server/src/db/lifecycle.rs:295]: My IDE was complaining the order of definition didn't match the order of the trait, so I moved this to pacify it

**Review by pauldix — COMMENTED** [review]
What counts as an access?

**Comment by tustvold** [comment]
@pauldix writes and queries that hit the underlying storage (i.e. aren't pruned out)

File: lifecycle/src/policy.rs [hunk 2]
```diff
@@ -1187,17 +1159,17 @@ mod tests {
         // Should first unload, then drop
         //
         // Should order the same actions by write time, with nulls last
```
Comment (alamb) [inline lifecycle/src/policy.rs:1161]:
```suggestion
        // Should order the same actions by access time, with nulls last
```

**Review by alamb — APPROVED** [review]
I think this PR does what the description says (orders by last access). 

I also reviewed the test logic changes carefully and it looks good to me 👍

**Pull request merged**